docs/USAGE-GUIDE.md

```diff
@@ -1777,6 +1777,210 @@ handle. The handle is considered invalid after **s2n_connection_free** is used.
+
+**s2n_connection_set_psk_mode** sets the PSK mode on the connection object. The enum **s2n_psk_mode** lists the supported PSK modes. This API overrides the PSK mode set on config for this connection. If the connection already contains a list of PSKs **s2n_connection_set_psk_mode** verifies that the current PSK mode matches that of the list. If the PSK mode being set differs from the existing list, an error is returned. If there is no existing list of PSKs present on the connection, the PSK mode of the list is set to the input PSK mode.
+
+### Selecting a Pre-Shared Key
+
+By default, s2n server chooses the first identity in its PSK list that also appears in the client's PSK list. If you would like to implement your own PSK selection logic,
+use the **s2n_psk_selection_callback** to select the PSK to be used for the connection, along with the following offered PSK APIs to process the client sent list of PSKs.
+
+```c
+struct s2n_offered_psk* s2n_offered_psk_new();
+int s2n_offered_psk_free(struct s2n_offered_psk **psk);
+```
+
+**s2n_offered_psk** creates a new offered PSK object. An offered PSK represents a single PSK offered by the client.
+Use this object to create the list of offered PSKs sent by the client. Use **s2n_offered_psk_free** to free
+the memory allocated to the offered PSK object created by this API.
+
+**s2n_offered_psk_free** frees the memory associated with the **s2n_offered_psk** object.
+
+The server calls the following APIs to select the PSK to be used for the connection.
+
+```c
+int s2n_offered_psk_get_identity(struct s2n_offered_psk *psk, uint8_t** identity, uint16_t *size);
+bool s2n_offered_psk_list_has_next(struct s2n_offered_psk_list *psk_list);
+int s2n_offered_psk_list_next(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+int s2n_offered_psk_list_reread(struct s2n_offered_psk_list *psk_list);
+int s2n_offered_psk_list_choose_psk(struct s2n_offered_psk_list *psk_list, struct s2n_offered_psk *psk);
+typedef int (*s2n_psk_selection_callback)(struct s2n_connection *conn, void *context,
+                                          struct s2n_offered_psk_list *psk_list);
+int s2n_config_set_psk_selection_callback(struct s2n_config *config, s2n_psk_selection_callback cb, void *context);
+```
+
+Use the **s2n_config_set_psk_selection_callback** only when a custom PSK selection logic is desired by the client.
+By default, s2n-tls server chooses the first identity in its PSK list that also appears in the client's PSK list.
+
+**s2n_offered_psk_get_identity** gets the PSK identity and PSK identity length for a given offered PSK object.
+Callers can use this API to obtain the identity value set on the **s2n_offered_psk** object.
+
```

Comment:
This API returns a pointer to internal memory with a limited lifetime. You don't discuss safety in [the doxygen comment](https://github.com/aws/s2n-tls/blob/main/api/s2n.h#L866-L874). You could discuss safety here, but it would probably make more sense in the doxygen comment.

Reply:
The lifetime of the identity is till the end of the PSK selection callback.